### PR TITLE
add support for retain_graph in backward

### PR DESCRIPTION
**Pull request by davidgonmar**
Prerequisite for https://github.com/tinygrad/tinygrad/issues/5858.
Simply allows not deleting the graph information when running backward (defaults to False). Most common usage is gradient accumulation or second order derivatives. Defaults to False as PyTorch.

Another important usage as I mentioned is gradient accumulation (you can look it up, it is a simple concept). See the test for the expected behaviour (to match pytorch). In order for it to be correct, the gradient accumulation needs to only be done on leaf tensors (hence the t0.grad = None in .backward(). In eager frameworks like PyTorch, this reduces memory usage significantly. It could also reduce it on Tinygrad for training (but I'm not 100% sure it will). It simply frees grads that will not be used).

### CODE REVIEW

**Comment by davidgonmar** [comment]
Update:
when you do this in torch
```
import torch
import numpy as np

# Define the size N
N = 10  # You can change N to the desired value

# Create a tensor with shape (N*N,) filled with 60000, half-precision, and requires gradient
t = torch.tensor([60000] * (N * N), dtype=torch.float16, requires_grad=True).reshape(N, N)

# Check if the mean of the tensor is 60000
np.testing.assert_allclose(t.mean().item(), 60000)

# Compute the mean of the square of the tensor and perform backpropagation
t.square().mean().backward()

# Check the gradients
np.testing.assert_allclose(t.grad.numpy().flatten(), [60000 * 2 / (N * N)] * (N * N))
```
it will throw since t.grad is None. Why? because t is not a leaf. In tinygrad, a lot of tests rely on the fact that intermediate gradients are kept. This will probably increase memory usage in those cases since intermediate tensors that are present in the Python scope will 'maintain' their grads, even though they are useless. What PyTorch does is t.grad = None just after calling backward. The only grads that are kept are the ones for leaf tensors. I think what PyTorch does is more 'scalable'. In order to emulate it, what I did is, after calling .backward a second time (assuming the tensor graph is kept), before running the actual .backward, I set all non-leaf graphs to None. This is necessary to allow gradient accumulation and second order derivatives (from the user perspective). Ideally, I think this should be changed to match the pytorch behaviour (and some tests changed not to expect grads on non-leaf tensors). If this seems good, I can refactor everything later after merging this.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -737,23 +737,26 @@ def multinomial(self:Tensor, num_samples:int = 1, replacement:bool = False) -> T
...
     If the 'gradient' argument is not provided, the tensor must be a scalar, and the gradient is implicitly set to 1.0.
-
+    If 'retain_graph' is false, the graph used to compute the grads will be freed. Otherwise, it will be kept. Keeping it can increase memory usage.
     ```python exec="true" source="above" session="tensor" result="python"
     t = Tensor([1.0, 2.0, 3.0, 4.0], requires_grad=True)
     t.sum().backward()
     print(t.grad.numpy())
     ```
     """
+    toposorted = self._deepwalk()
```
Comment (geohot) [inline tinygrad/tensor.py:759]: Why does this have to move?

Comment (davidgonmar) [reply tinygrad/tensor.py:759]: > Why does this have to move?

On the 'Update comment' I mentioned the details of what I want to achieve. I need to call the deepwalk before assigning the gradient to the root tensor, since the deepwalk resets the grads. If I reset it after setting it, then the root grad is 'None' and that's not what we want.
